### PR TITLE
Fix GraphQL Explorer

### LINKED ISSUE #15304: InvalidOperationException when visiting GraphQL explorer
I've just updated to the latest OC and getting this when visiting GraphiQL explorer:

```
System.InvalidOperationException: This graph type 'AutorouteInputObjectType' with name 'AutoroutePartInput' has already been initialized. Make sure that you do not use the same instance of a graph type in multiple schemas. It may be so if you registered this graph type as singleton; see https://graphql-dotnet.github.io/docs/getting-started/dependency-injection/ for more info.
   at GraphQL.Types.GraphType.Initialize(ISchema schema) in /_/src/GraphQL/Types/GraphType.cs:line 44
   at GraphQL.Types.SchemaTypes.AddType(IGraphType type, TypeCollectionContext context) in /_/src/GraphQL/Types/Collections/SchemaTypes.cs:line 522
   at GraphQL.Types.SchemaTypes.AddTypeIfNotRegistered(IGraphType type, TypeCollectionContext context) in /_/src/GraphQL/Types/Collections/SchemaTypes.cs:line 776
   at GraphQL.Types.SchemaTypes.Initialize(ISchema schema, IServiceProvider serviceProvider, IEnumerable`1 graphTypeMappings) in /_/src/GraphQL/Types/Collections/SchemaTypes.cs:line 277
   at GraphQL.Types.SchemaTypes..ctor(ISchema schema, IServiceProvider serviceProvider, IEnumerable`1 graphTypeMappings) in /_/src/GraphQL/Types/Collections/SchemaTypes.cs:line 190
   at GraphQL.Types.SchemaTypes..ctor(ISchema schema, IServiceProvider serviceProvider) in /_/src/GraphQL/Types/Collections/SchemaTypes.cs:line 175
   at GraphQL.Types.Schema.CreateSchemaTypes() in /_/src/GraphQL/Types/Schema.cs:line 477
   at GraphQL.Types.Schema.CreateAndInitializeSchemaTypes() in /_/src/GraphQL/Types/Schema.cs:line 448
   at GraphQL.Types.Schema.Initialize() in /_/src/GraphQL/Types/Schema.cs:line 196
   at OrchardCore.Apis.GraphQL.Services.SchemaService.GetSchemaAsync() in C:\repo\OrchardCore\src\OrchardCore.Modules\OrchardCore.Apis.GraphQL\Services\SchemaService.cs:line 116
   at OrchardCore.Apis.GraphQL.GraphQLMiddleware.ExecuteAsync(HttpContext context) in C:\repo\OrchardCore\src\OrchardCore.Modules\OrchardCore.Apis.GraphQL\GraphQLMiddleware.cs:line 142
   at OrchardCore.Apis.GraphQL.GraphQLMiddleware.InvokeAsync(HttpContext context, RequestDelegate next) in C:\repo\OrchardCore\src\OrchardCore.Modules\OrchardCore.Apis.GraphQL\GraphQLMiddleware.cs:line 65
   at Microsoft.AspNetCore.Builder.UseMiddlewareExtensions.InterfaceMiddlewareBinder.<>c__DisplayClass2_0.<<CreateMiddleware>b__0>d.MoveNext()
```

image

### PR DESCRIPTION
**Pull request by MikeAlhayek**
Fix #15304

@lampersky can you please checkout this branch and confirm that it fixes your issues?

The exception states 

> System.InvalidOperationException: 'This graph type 'AliasInputObjectType' with name 'AliasPartInput' has already been initialized. Make sure that you do not use the same instance of a graph type in multiple schemas. It may be so if you registered this graph type as singleton; see https://graphql-dotnet.github.io/docs/getting-started/dependency-injection/ for more info.'

### CODE REVIEW

**Comment by MikeAlhayek** [comment]
@hyzx86 can you please review the changes in this PR?

**Comment by sebastienros** [comment]
I did this change I believe (not even sure but seems familiar). And the comments states it is singleton, so the comment could be removed "if this is the correct thing to do". Have you tried to understand what the issue was with the resolution? It really can't be a singleton?

**Comment by MikeAlhayek** [comment]
@sebastienros I updated the description in the PR to show the detailed in the exception. I am guessing after the upgrade, we no longer need it to be singleton. And after the change everything seems to be working as expected. These types are resolved once during the schema building.

File: src/OrchardCore/OrchardCore.ContentManagement.GraphQL/Queries/Types/TypedContentTypeBuilder.cs [hunk 1]
```diff
@@ -57,13 +58,14 @@ public void Build(FieldType contentQuery, ContentTypeDefinition contentTypeDefin
...
-                                                                    {
-                                                                        queryObjectGraphTypes ??= serviceProvider.GetService<IEnumerable<IObjectGraphType>>();
-                                                                        return queryObjectGraphTypes?.FirstOrDefault(x => x.GetType().BaseType.GetGenericArguments().First().Name == partName);
-                                                                    }
-                                                                );
+                var queryGraphType = _partObjectGraphTypes
+                    .GetOrAdd(part.PartDefinition.Name,
+                        partName =>
+                        {
+                            queryObjectGraphTypes ??= serviceProvider.GetServices<IObjectGraphType>();
```
Comment (sebastienros) [inline src/OrchardCore/OrchardCore.ContentManagement.GraphQL/Queries/Types/TypedContentTypeBuilder.cs:65]: The problem might be here. `seviceProvider` is stored in the closure. It should be resolved within this method and not outside.

Comment (MikeAlhayek) [reply src/OrchardCore/OrchardCore.ContentManagement.GraphQL/Queries/Types/TypedContentTypeBuilder.cs:65]: I think we should not store the result of the queryObject outside that method. Yes. But that did not solve the issue.

image

File: src/OrchardCore/OrchardCore.ContentManagement.GraphQL/Queries/Types/TypedContentTypeBuilder.cs [hunk 2]
```diff
@@ -62,7 +58,7 @@ public void Build(FieldType contentQuery, ContentTypeDefinition contentTypeDefin
                     .GetOrAdd(part.PartDefinition.Name,
                         partName =>
                         {
-                            queryObjectGraphTypes ??= serviceProvider.GetServices<IObjectGraphType>();
+                            var queryObjectGraphTypes = serviceProvider.GetServices<IObjectGraphType>();
```
Comment (sebastienros) [inline src/OrchardCore/OrchardCore.ContentManagement.GraphQL/Queries/Types/TypedContentTypeBuilder.cs:61]: I was wrong with the service provider, I thought the lambda was long-lived. But the change you made might now trigger multiple resolutions of the same thing. It was ok as the variable was in the method.

Comment (MikeAlhayek) [reply src/OrchardCore/OrchardCore.ContentManagement.GraphQL/Queries/Types/TypedContentTypeBuilder.cs:61]: I don't think it trigger multiple times. The types are registered scooped and only created once when the app starts and the first request is made to the GraphiQL middleware

Comment (MikeAlhayek) [reply src/OrchardCore/OrchardCore.ContentManagement.GraphQL/Queries/Types/TypedContentTypeBuilder.cs:61]: @sebastienros 
At the same time, I do not see a good reason to resolve them in the callback. Why can't we resolve them early out out side the loop? like this

```
            var queryObjectGraphTypes = serviceProvider.GetServices<IObjectGraphType>();
            var queryInputGraphTypes = serviceProvider.GetServices<IInputObjectGraphType>();
```

File: src/OrchardCore/OrchardCore.Apis.GraphQL.Abstractions/ServiceCollectionExtensions.cs [hunk 3]
```diff
@@ -34,9 +32,9 @@ public static class ServiceCollectionExtensions
         {
             // Instances are registered as singletons as their constructor holds the logic to configure the type
             // and doesn't need to run every time
-            services.AddSingleton<TInputType>();
-            services.AddSingleton<ObjectGraphType<TInput>, TInputType>(s => s.GetRequiredService<TInputType>());
-            services.AddSingleton<IObjectGraphType, TInputType>(s => s.GetRequiredService<TInputType>());
+            services.AddScoped<TInputType>();
```
Comment (sebastienros) [inline src/OrchardCore/OrchardCore.Apis.GraphQL.Abstractions/ServiceCollectionExtensions.cs:35]: These can be transient too

File: src/OrchardCore/OrchardCore.Apis.GraphQL.Abstractions/ServiceCollectionExtensions.cs [hunk 4]
```diff
@@ -34,9 +32,9 @@ public static class ServiceCollectionExtensions
         {
             // Instances are registered as singletons as their constructor holds the logic to configure the type
```
Comment (sebastienros) [inline src/OrchardCore/OrchardCore.Apis.GraphQL.Abstractions/ServiceCollectionExtensions.cs:33]: Remove comment

**Review by sebastienros — APPROVED** [review]

**Comment by sebastienros** [comment]
Tried on my side and came to the same solution, using Transient though. Each type shouldn't be resolved multiple time during a request that builds the schema.

**Comment by lampersky** [comment]
@MikeAlhayek, @sebastienros I've just tried `ma/fix-graphql` branch, and can't repro the issue anymore, so it looks the fix is solving the issue. Scheme is reloaded properly and GraphiQL explorer works as expected.

I only can see plenty of exceptions, when navigating from `Admin/GraphQL` to `/Admin/Contents/ContentItems`, but this is not related to issue reported earlier:
image

**Comment by gvkries** [comment]
Note that by making the graph types `Scoped`, it is not possible to resolve them automatically while building the schema. That means using something like 

```
Field<[GraphType]>("Foo")
  .Description("bar")
  .ResolveAsync(x =>
  {
    [...]
  });
```
fails for complex types, because the  `ISchemaFactory` is a singleton and a scoped `[GraphType]` cannot be resolved.  This happens when a graph type depends on another one. 
Also note that any scoped instance may still be referenced by the schema.

**Pull request merged**